### PR TITLE
add minifollowups v1

### PR DESCRIPTION
**Pull request by ahnitz**
This adds followup plots for the top X foreground events. It's at the point that I think it is worth recommitting to master. 

The following features are implemented
* trigger time series plots
* snr/chisq plots (created from a pycbc_inspiral based single template followup code)
* tables with the easiest to extrac coinc / single trigger information. 

If need be I can split this into separate pull requests for each script, followed by the changes to the workflow. 

Example page using the latest code.
https://sugar-jobs.phy.syr.edu/~ahnitz/projects/bigdog_rerun/plots/gw12/html/7._result/

### CODE REVIEW

**Comment by cmbiwer** [comment]
This line is going to cause an error: https://github.com/ahnitz/pycbc/blob/mini/pycbc/workflow/minifollowups.py#L42

As you removed the ``output_filelist`` initialization.

**Comment by ahnitz** [comment]
Thanks Chris, fixed.

**Comment by ahnitz** [comment]
Ian, I am happy enough with the state of the branch now, cany you give it a look over and give me your feedback. I'll post some ini file updates for the ER8 run, if we choose to use this.

File: bin/hdfcoinc/pycbc_make_hdf_coinc_workflow [hunk 1]
```diff
@@ -197,7 +192,7 @@ for insp in full_insps:
     hist_summ += list(grouper(hists, 2))
 
 # Calculate the inspiral psds and make plots
-full_segs, psd_files, insp_segs, insp_data_segs = [], [], {}, {}                            
+full_segs, psd_files, insp_segs, insp_data_segs = [], [], {}, {}
```
Comment (spxiwh) [inline bin/hdfcoinc/pycbc_make_hdf_coinc_workflow]: Why did you only remove some of the excess white space?

File: bin/minifollowups/pycbc_page_coincinfo [hunk 2]
```diff
@@ -0,0 +1,97 @@
+#!/bin/env python
```
Comment (spxiwh) [inline bin/minifollowups/pycbc_page_coincinfo]: Please insert the normal copyright mumbo jumbo

File: bin/minifollowups/pycbc_page_coincinfo [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+args = parser.parse_args()
+pycbc.init_logging(args.verbose)
+
+# Get the nth loudest trigger from the output of pycbc_coinc_statmap
+f = h5py.File(args.statmap_file, 'r')
+try:
+    n = f['foreground/stat'][:].argsort()[::-1][args.n_loudest]
+except:
+    pylab.text(0.5, 0.5, 'no trigger found'); pylab.savefig(args.output_file)
```
Comment (spxiwh) [inline bin/minifollowups/pycbc_page_coincinfo]: "No triggers found"

File: bin/hdfcoinc/pycbc_make_hdf_coinc_workflow (hunk 1 above)
Comment (ahnitz) [inline bin/hdfcoinc/pycbc_make_hdf_coinc_workflow]: I think not intentionally. I originally though I should rewrite this line to include some other information, but decided it was simpler to leave as is and so reverted it, with apparently a whitespace change.

File: bin/minifollowups/pycbc_plot_trigger_timeseries [hunk 4]
```diff
@@ -0,0 +1,97 @@
+#!/bin/env python
```
Comment (spxiwh) [inline bin/minifollowups/pycbc_plot_trigger_timeseries]: Copyright words

File: bin/minifollowups/pycbc_page_coincinfo (hunk 2 above)
Comment (ahnitz) [inline bin/minifollowups/pycbc_page_coincinfo]: ok, will do

File: bin/minifollowups/pycbc_plot_trigger_timeseries [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+def get_time_from_cli(args):
+    """ Return the time and single detector trigger ids for the nth loudest
+    coincident trigger """
+    if args.statmap_file:
+        f = h5py.File(args.statmap_file, 'r')
+        try:
+            stat = f['foreground/stat'][:].argsort()[::-1][args.n_loudest]
+        except:
+            pylab.text(0.5, 0.5, 'no trigger found'); pylab.savefig(args.output_file)
```
Comment (spxiwh) [inline bin/minifollowups/pycbc_plot_trigger_timeseries]: triggers

File: bin/minifollowups/pycbc_single_template_plot [hunk 6]
```diff
@@ -0,0 +1,59 @@
+#!/bin/env python
```
Comment (spxiwh) [inline bin/minifollowups/pycbc_single_template_plot]: Copyright words

File: bin/minifollowups/pycbc_single_template_plot [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+# The event is chosen from the input of pycbc_single template, so we
+# just extract the parameters here
+f = h5py.File(args.single_template_file)
+
+try:
+    delta_t = f['snr'].attrs['delta_t']
+    start_time = f['snr'].attrs['start_time']
+    time = f.attrs['event_time']
+except:
+    pylab.text(0.5, 0.5, 'no trigger found'); pylab.savefig(args.output_file)
```
Comment (spxiwh) [inline bin/minifollowups/pycbc_single_template_plot]: triggers

File: bin/pycbc_single_template [hunk 8]
```diff
@@ -1,8 +1,34 @@
 #!/usr/bin/env python
```
Comment (spxiwh) [inline bin/pycbc_single_template]: Copyright words

File: bin/pycbc_single_template [hunk 9]
```diff
@@ -1,8 +1,34 @@
...
-import sys, logging, argparse, numpy, itertools, pycbc
+""" Calculate the SNR and CHIQ timeseries for either a chosen template, or
```
Comment (spxiwh) [inline bin/pycbc_single_template]: CHIQ -> CHISQ

File: pycbc/workflow/minifollowups.py [hunk 10]
```diff
@@ -15,76 +15,146 @@
...
-                             coinc_file, tmpltbank_file, data_type, tags=None):
-    ''' This performs a series of followup jobs on the num_events-th loudest
-    events.
-    '''
-
+from pycbc.workflow.core import Executable, FileList, Node, makedir
+from pycbc.workflow.plotting import PlotExecutable, requirestr, excludestr
+from itertools import izip_longest
+
+def grouper(iterable, n, fillvalue=None):
```
Comment (spxiwh) [inline pycbc/workflow/minifollowups.py]: Can you add documentation, as I am unsure what this function actually does!

**Comment by spxiwh** [comment]
Hi Alex,

Thanks for posting this. I made some in-line feedback. Once that is addressed and checks completed, please merge.

**Comment by ahnitz** [comment]
Ok, thanks Ian. I think I've fixed all the things you've found.

**Pull request merged**